Sources/Customization/DefaultEventProcessor.swift

```diff
@@ -0,0 +1,271 @@
+                } else {
+                    self.logger.e("Failed to removed \(num) events")
+                }
+            }
+            
+            while let eventsToSend: [EventForDispatch] = self.dataStore.getFirstItems(count: self.batchSize) {
+                let (numEvents, batched) = eventsToSend.batch()
+                
+                guard numEvents > 0 else { break }
+                
+                guard let batchEvent = batched else {
+                    // discard an invalid event that causes batching failure
+                    // - if an invalid event is found while batching, it batches all the valid ones before the invalid one and sends it out.
+                    // - when trying to batch next, it finds the invalid one at the header. It discards that specific invalid one and continue batching next ones.
+
+                    removeStoredEvents(num: 1)
+                    continue
+                }
+                
+                // we've exhuasted our failure count.  Give up and try the next time a event
+                // is queued or someone calls flush.
+                if failureCount > DefaultValues.maxFailureCount {
+                    self.logger.e(.eventSendRetyFailed(failureCount))
+                    break
+                }
+                
+                // send notification BEFORE sending event to the server
+                NotificationCenter.default.post(name: .willSendOptimizelyEvents, object: batchEvent)
+                
+                // make the send event synchronous. enter our notify
+                self.notify.enter()
+                self.eventHandler.dispatch(event: batchEvent) { (result) -> Void in
+                    switch result {
+                    case .failure(let error):
+                        self.logger.e(error.reason)
+                        failureCount += 1
+                    case .success:
+                        // we succeeded. remove the batch size sent.
+                        removeStoredEvents(num: numEvents)
+
+                        // reset failureCount
+                        failureCount = 0
+                    }
+                    // our send is done.
+                    self.notify.leave()
+                    
+                }
+                // wait for send
+                self.notify.wait()
+            }
+        }
+    }
+    
+    func applicationDidEnterBackground() {
+        stopTimer()
+        
+        flush()
+    }
+    
+    func applicationDidBecomeActive() {
+        if dataStore.count > 0 {
+            startTimer()
+        }
+    }
+    
+    func startTimer() {
+        // timer is activated only for iOS10+ and non-zero interval value
+        guard #available(iOS 10.0, tvOS 10.0, *), timerInterval > 0 else {
+            flush()
+            return
+        }
+        
+        guard self.timer.property == nil else { return }
+        
+        DispatchQueue.main.async {
```

Comment:
Why do we use `DispatchQueue.main` for the timer task, but our own private `DispatchQueue` for flushing?

Reply:
in that async call you are setting a timer which is part of the main event loop.  thus it must be done on the main thread.  we had considered moving this to a dispatch timer but felt this was right for background foreground.  if you attempt a timer call in the background it will crash as backgrounding takes a delegate pattern as opposed to callback.